Story/getting stared revision (#109)
vmware/vmware-aria-operations-integration-sdk: README.md

> The VMware Aria Operations Integration SDK has been tested in the following OSes:

Reviewer: OSes -> OSs

The VMware Aria Operations Integration SDK has been tested in the following OSes:
* Windows 10
* Windows 11
* macOS 12 (Monterey)
* macOS 13 (Ventura)
* Debian Linux
* Fedora Linux

Other operating systems may be compatible.

#### VMware Aria Operations

> The Management Packs generated by the VMware Aria Operations Integration SDK will only run on versions that supports containerized Management Packs. Currently, this is limited to version 8.10 or later.

Reviewer 2: This line shouldn't change. We're still on-prem only for external users.

   The icon is used in the VMware Aria Operations UI if present. If it is not present, a default icon will be used. The icon file must be
   PNG format and 256x256 pixels. An icon file can be added later by copying the icon to the root project directory and
   setting the value of the `"pak_icon"` key to the icon's file name in the `manifest.txt` file.

[//]: # (TODO: Add this section back when we support it)
[//]: # (8. `Select a language for the adapter`)
[//]: # (  selected language. The template adapter collects several objects and metrics from the container that the adapter)
[//]: # (  Once selected, the project will be generated, including a template adapter in the)
[//]: # (  is running in, and can be used as a starting point for creating a new adapter.)

For complete documentation of the `mp-init` tool including an overview of its output, see the [MP Initialization Tool Documentation](doc/mp-init.md).

### Template Adapter
The template adapter has comments throughout the code to help new users understand the process of creating their own adapter using the existing code. For additional guidance creating adapters see the [Walkthroughs](../README.md#walkthroughs) section.

> The code inside `adapter.py`  has four methods:

Reviewer 2: This section starts talking about the template adapter right away without any discussion about that the template adapter is or where it's located or why we should care. Also, we should be careful about talking about things that are specific to the template adapter vs things that are generally true of any adapter. The methods below are in the latter category, so talking about them only in terms of the sample adapter is potentially misleading.

7. `Enter a path to the Management Pack icon file, or leave blank for no icon`

   The icon is used in the VMware Aria Operations UI if present. If it is not present, a default icon will be used. The icon file must be
   PNG format and 256x256 pixels. An icon file can be added later by copying the icon to the root project directory and
   setting the value of the `"pak_icon"` key to the icon's file name in the `manifest.txt` file.

[//]: # (TODO: Add this section back when we support it)
[//]: # (8. `Select a language for the adapter`)
[//]: # (  selected language. The template adapter collects several objects and metrics from the container that the adapter)
[//]: # (  Once selected, the project will be generated, including a template adapter in the)
[//]: # (  is running in, and can be used as a starting point for creating a new adapter.)

For complete documentation of the `mp-init` tool including an overview of its output, see the [MP Initialization Tool Documentation](doc/mp-init.md).

### Template Adapter

> The template adapter has comments throughout the code to help new users understand the process of creating their own adapter using the existing code. For additional guidance creating adapters see our [Walkthroughs](../README.md#walkthroughs) section.

Reviewer 2: ```suggestion
The template adapter has comments throughout the code to help new users understand the process of creating their own adapter using the existing code. For additional guidance creating adapters see the [Walkthroughs](../README.md#walkthroughs) section.
```

> The code inside `adapter.py` executed by an HTTP sever that runs inside container every time a test collection, or a collection is made[^1]. The HTTP server sends the information returned by the executable back to VMware Aria Operations. the template executable, `adapter.py` has four methods:

Reviewer 2: [^1] doesn't render the way I assume you want it to. Regardless, I don't think we should mention that the adapter is stateless in a footnote. I think this could use some cleaning up in general - some bits are confusing or not relevant at this point. We should talk about the architecture, but not here.

- test(adapter_instance):
Performs a test connection using the information given to the adapter_instance to verify the adapter has been set up properly. A typical test connection will generally consist of:

 1. Read identifier values from adapter_instance that are required to connect to the target(s)
 2. Connect to the target(s), and retrieve some sample data
 3. Disconnect cleanly from the target (ensure this happens even if an error occurs)
 4. If any of the above failed, return an error, otherwise pass.

- get_endpoints(adapter_instance):
This  method will be run before the 'test' method, and VMware Aria Operations will use
the results to extract a certificate from each URL. If the certificate is not trusted by
the VMware Aria Operations Trust Store, the user will be prompted to either accept or reject
the certificate. If it is accepted, the certificate will be added to the AdapterInstance
object that is passed to the 'test' and 'collect' methods. Any certificate that is
encountered in those methods should then be validated against the certificate(s)

> in the AdapterInstance.

Reviewer 2: We should mention that this is only applicable for HTTPS endpoints. If we have a different type (e.g., a database connection), it will not work, and there is no benefit to returning anything here.

- collect(adapter_instance):
Performs a collection against the target host. A typical collection will generally consist of:
1. Read identifier values from adapter_instance that are required to connect to the target(s)
2. Connect to the target(s), and retrieve data
3. Add the data into a CollectResult's objects, properties, metrics, etc
4. Disconnect cleanly from the target (ensure this happens even if an error occurs)
5. Return the CollectResult.

- get_adapter_definition():
Defines the object types and attribute types (metric/property) that are present
in a collection. Setting these object types and attribute types helps VMware Aria Operations to

> validate, process, and display the data correctly.

Reviewer 2: More importantly in most cases, it defines the Adapter Instance configuration parameters/credentials.

> Each method represents a single request, and it can be tested individually using `mp-test`, which is covered in the section ahead.

Reviewer 2: Each method represents a single request, and it can be tested individually using `mp-test` (see section [...]).

> [^1]The HTTP server does not record the state inside the executable files, so the adapter should be considered stateless.

Reviewer 2: This should be higher and reworded. We don't need to mention the HTTP server or 'executable files'. We should word the part about being stateless stronger; namely that it _is_ stateless, not that it should be _considered_ stateless. We can mention how it is run in a stateless way: That each request starts its own python subprocess which runs the appropriate method and then exits.

   PNG format and 256x256 pixels. An icon file can be added later by copying the icon to the root project directory and
   setting the value of the `"pak_icon"` key to the icon's file name in the `manifest.txt` file.

[//]: # (TODO: Add this section back when we support it)
[//]: # (8. `Select a language for the adapter`)
[//]: # (  selected language. The template adapter collects several objects and metrics from the container that the adapter)
[//]: # (  Once selected, the project will be generated, including a template adapter in the)
[//]: # (  is running in, and can be used as a starting point for creating a new adapter.)

For complete documentation of the `mp-init` tool including an overview of its output, see the [MP Initialization Tool Documentation](doc/mp-init.md).

### Template Project
Every new project creates a file system that has the basic project structure required to develop and build and Management Pack.
Each file and directory is discussed in depth in the [mp-init](doc/mp-init.md) documentation. `app/adapter.py` is the entry point
for the your adapter, and is the best starting point. `adapter.py` is a template

> adapter that [what does it do?]. It has comments throughout its code that explain what the code in the template adapter does and how to turn it into your own adapter. The methods inside the adapter template are required and should not be deleted; however, the code inside the methods should be modified to generate the desired adapter. Each method is used for a different function as described below:

Reviewer 2: `[what does it do?]` is a placeholder - make sure you fill it in

> Every new project creates a file system that has the basic project structure required to develop and build and Management Pack.
> Each file and directory is discussed in depth in the [mp-init](doc/mp-init.md) documentation. `app/adapter.py` is the entry point
> for the your adapter, and the best starting point for a new user to begin their development journey. `adapter.py` is a template
> adapter that has comments throughout its code that allow users to replace the code in the template adapter and turn it into their own adapter. The methods inside the adapter template are all required and should not be deleted; however, the code inside the methods should be substituted to generate the desired adapter. Each method is used for a different function as described below:

Reviewer 2: ```suggestion
Every new project creates a file system that has the basic project structure required to develop and build and Management Pack.
Each file and directory is discussed in depth in the [mp-init](doc/mp-init.md) documentation. `app/adapter.py` is the entry point
for the your adapter, and is the best starting point. `adapter.py` is a template
adapter that [what does it do?]. It has comments throughout its code that explain what the code in the template adapter does and how to turn it into your own adapter. The methods inside the adapter template are required and should not be deleted; however, the code inside the methods should be modified to generate the desired adapter. Each method is used for a different function as described below:
```

This reads much better than what was there before! I made some small tweaks and a section to fill in. The main thing that I noticed is you were switching between second person (e.g., "your adapter") third person (e.g., "a new user").

Also, while I appreciate the focus on new users, in many cases the same things will apply to more experienced users as well (e.g., starting with 'adapter.py'), so I'd recommend against being more specific than necessary.

Reviewer 2: Also, 'get_adapter_definition' is optional (but recommended - if it's omitted there should be a manually-created `describe.xml` file instead)

the results to extract a certificate from each URL. If the certificate is not trusted by
the VMware Aria Operations Trust Store, the user will be prompted to either accept or reject
the certificate. If it is accepted, the certificate will be added to the AdapterInstance
object that is passed to the 'test' and 'collect' methods. Any certificate that is
encountered in those methods should then be validated against the certificate(s)
in the AdapterInstance.

- collect(adapter_instance):
Performs a collection against the target host. A typical collection will generally consist of:
1. Read identifier values from adapter_instance that are required to connect to the target(s)
2. Connect to the target(s), and retrieve data
3. Add the data into a CollectResult's objects, properties, metrics, etc
4. Disconnect cleanly from the target (ensure this happens even if an error occurs)
5. Return the CollectResult.

- get_adapter_definition():
Defines the object types and attribute types (metric/property) that are present
in a collection. Setting these object types and attribute types helps VMware Aria Operations to
validate, process, and display the data correctly.

> Each method represents a single request, and it can be tested individually using `mp-test`, which is covered in the following section. The adapter is stateless,therefore the test __test__ and __collect__ methods should be self contained. For further guidance on using the template project consult the [Walkthroughs](../README.md#walkthroughs) section.

Reviewer 2: * I think this should be before the descriptions of the methods
* "self-contained" is probably not the right word, since they can certainly call other methods, libraries, files, etc. Also none of the other methods can store state either, not just the test and collect methods.

Maybe something like this:
```suggestion
Each method represents a single request, and it can be tested individually using `mp-test`, which is covered in the following section. The adapter is stateless, therefore the methods are not able to store any data for use in later method calls. For further guidance on using the template project consult the [Walkthroughs](../README.md#walkthroughs) section.
```

[//]: # (  is running in, and can be used as a starting point for creating a new adapter.)

For complete documentation of the `mp-init` tool including an overview of its output, see the [MP Initialization Tool Documentation](doc/mp-init.md).

### Template Project
Every new project creates a file system that has the basic project structure required to develop and build a Management Pack.
Each file and directory are discussed in depth in the [mp-init](doc/mp-init.md) documentation. `app/adapter.py` is the adapter's
entry point and the best starting point. `adapter.py` is a template adapter that collects several objects and metrics from the
container in which the adapter is running; use the template as a starting point for creating a new adapter. The template adapter 
has comments throughout its code that explain what the code does and how to turn it into your adapter. The methods inside the adapter
template are required. Modify the code inside the methods to generate the desired adapter. Each method represents a single request,
and it can be tested individually using `mp-test`, which is covered in the following section. The adapter is stateless; therefore,

> the methods don't store any data for use in later method calls. Each method is used for a different function as described below:

Reviewer 2: ```suggestion
the adapter cannot store any data for use in later method calls. Each method is used for a different function as described below:
```

- test(adapter_instance):

>   Performs a test connection using the information given to the adapter_instance to verify the adapter has been set up properly.

Reviewer 2: ```suggestion
  Performs a test connection using the information given to the adapter_instance to verify the adapter instance has been configured properly.
```

  A typical test connection will generally consist of:

     1. Read identifier values from adapter_instance that are required to connect to the target(s)
     2. Connect to the target(s), and retrieve some sample data

>      3. Disconnect cleanly from the target (ensure this happens even if an error occurs)
>      4. If any of the above failed, return an error, otherwise pass.

Reviewer 2: ```suggestion
     3. If any of the above failed, return an error, otherwise pass.
     4. Disconnect cleanly from the target (ensure this happens even if an error occurs)
```

  This method will be run before the 'test' method, and VMware Aria Operations will use
  the results to extract a certificate from each URL. If the certificate is not trusted by
  the VMware Aria Operations Trust Store, the user will be prompted to either accept or reject
  the certificate. If it is accepted, the certificate will be added to the AdapterInstance
  object that is passed to the 'test' and 'collect' methods. Any certificate that is
  encountered in those methods should then be validated against the certificate(s)
  in the AdapterInstance. This method will not only work against HTTPS endpoints, different types
  of endpoint will not work (eg. database connections).

- collect(adapter_instance):
  Performs a collection against the target host. A typical collection will generally consist of:
    1. Read identifier values from adapter_instance that are required to connect to the target(s)
    2. Connect to the target(s), and retrieve data
    3. Add the data into a CollectResult's objects, properties, metrics, etc
    4. Disconnect cleanly from the target (ensure this happens even if an error occurs)
    5. Return the CollectResult.

- get_adapter_definition():

>   Defines the object types and attribute types, and in most cases, it also defines the Adapter Instance configuration 
>   parameters/credentials present in a collection. Setting these helps VMware Aria Operations to validate, process, and display the 
>   data correctly.

Reviewer 2: ```suggestion
  Optional method that defines the Adapter Instance configuration (parameters and credentials used to connect to the target, and configure the management pack) present in a collection, and defines the object types and attribute types present in a collection. Setting these helps VMware Aria Operations to validate, process, and display the data correctly. If this method is omitted, a `describe.xml` file should be manually created inside the `conf` directory with the same data. Generally, this is only necessary when using advanced features of the `describe.xml` file that are not present in this method.
```

encountered in those methods should then be validated against the certificate(s)
in the AdapterInstance. This method will not only work against HTTPS endpoints, different types
of endpoint will not work and there is no benefit of returning them.

- collect(adapter_instance):
Performs a collection against the target host. A typical collection will generally consist of:
1. Read identifier values from adapter_instance that are required to connect to the target(s)
2. Connect to the target(s), and retrieve data
3. Add the data into a CollectResult's objects, properties, metrics, etc
4. Disconnect cleanly from the target (ensure this happens even if an error occurs)
5. Return the CollectResult.

- get_adapter_definition():
Defines the object types and attribute types (metric/property) that are present
in a collection. Setting these object types and attribute types helps VMware Aria Operations to
validate, process, and display the data correctly. More importantly, in most cases, it defines the
Adapter Instance configuration parameters/credentials.

Each method represents a single request, and it can be tested individually using `mp-test`, which is covered in the following section.

> The adapter is stateless, therefore the methods are not able to store any data for use in later method calls. For further guidance on

Reviewer 2: I think everything before 'For further guidance on...' should be in the intro section before the discussion on individual method calls

the certificate. If it is accepted, the certificate will be added to the AdapterInstance
object that is passed to the 'test' and 'collect' methods. Any certificate that is
encountered in those methods should then be validated against the certificate(s)
in the AdapterInstance. This method will not only work against HTTPS endpoints, different types

> of endpoint will not work and there is no benefit of returning them.

Reviewer 2: We should give examples of other types of endpoints

'no benefit of returning them' this seems redundant as we just said they won't work.

- get_adapter_definition():
Defines the object types and attribute types (metric/property) that are present
in a collection. Setting these object types and attribute types helps VMware Aria Operations to
validate, process, and display the data correctly. More importantly, in most cases, it defines the

> Adapter Instance configuration parameters/credentials.

Reviewer 2: If it's more important it should be the first thing we talk about.

'in most cases' -> if we include this phrase, there should be some discussion about what the different cases are and why they may be relevant. 

There should be at least a sentence or two about configuration parameters and credentials and how they're used in Aria Ops
